meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,435 @@
+import { TimelineBuilder } from './interfaces/timeline-builder'
+import { Rundown } from '../../model/entities/rundown'
+import {
+	ActivePartTimelineObjectGroup,
+	DeviceType,
+	Timeline,
+	TimelineObjectControl,
+	TimelineObjectGroup,
+	TimelineObjectType,
+} from '../../model/entities/timeline'
+import { TimelineObject } from '../../model/entities/timeline-object'
+import { Part } from '../../model/entities/part'
+import { Piece } from '../../model/entities/piece'
+import { TimelineEnable } from '../../model/entities/timeline-enable'
+import { TransitionType } from '../../model/enums/transition-type'
+import { PartTimings } from '../../model/value-objects/part-timings'
+import { PieceLifespan } from '../../model/enums/piece-lifespan'
+import { UnsupportedOperation } from '../../model/exceptions/unsupported-operation'
+
+const ACTIVE_GROUP_PREFIX: string = 'active_group_'
+const PREVIOUS_GROUP_PREFIX: string = 'previous_group_'
+const NEXT_GROUP_PREFIX: string = 'next_group_'
+const INFINITE_GROUP_PREFIX: string = 'infinite_group_'
+const PIECE_PRE_ROLL_PREFIX: string = 'pre_roll_'
+
+const PIECE_CONTROL_INFIX: string = '_piece_control_'
+const PIECE_GROUP_INFIX: string = '_piece_group_'
+
+// These priority values are the same values used by Core
+const HIGH_PRIORITY: number = 5
+const MEDIUM_PRIORITY: number = 1
+const LOW_PRIORITY: number = -1
+
+export class SuperflyTimelineBuilder implements TimelineBuilder {
+	public getBaseTimeline(): Timeline {
+		return { timelineGroups: [] }
+	}
+
+	public buildTimeline(rundown: Rundown): Timeline {
```

Comment:
This is going to sound like a cheeky answer, but as you point out there are more TODOs, so this isn't the final form of the method yet.

I did initially go with something like your proposed code snippet. I opted to not do that because your method `pustPreviousTimlineGroup` has "side effects". It modifies data outside of its scope (here it modifies `timeline`). That is not a good thing and we should try to avoid it when possible because "side effects" aren't always clear that they happen which would lead to "unexpected" behaviour.